### PR TITLE
Add WakeLock to Deck Import

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
User needs to keep tapping on their screen to keep the import going. This is annoying on large decks

## Fixes
Related #7077 

## Approach
Take a WakeLock

## How Has This Been Tested?

Untested, don't have enough space on my phone to test, and emulator doesn't sleep

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/async/WakeLockTaskListener.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import androidx.annotation.Nullable;
+import timber.log.Timber;
+
+import static android.os.PowerManager.*;
+
+@SuppressLint("DirectSystemCurrentTimeMillisUsage")
+public abstract class WakeLockTaskListener<TContext extends Context> extends TaskListenerWithContext<TContext> {
+
+    /* 30 minutes */
+    private static final long TIMEOUT_MS = 30 * 60 * 1000L;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/WakeLockTaskListener.java:37]: I don't think this is going to be long enough - Sync has no timeout, which is scary. 

This seemed like a reasonable middle-ground (with a silent exception if the timeout is exceeded), but I'm unsure

**Review by mikehardy — COMMENTED** [review]
Ooo - very interesting. I like the attempt / style you're going for here though, I see what you're saying with 30min maybe working but maybe not, with the exception report we'll see I guess.

Is there anything we can bubble up to the user to help them manage the process themselves? We can't know how long an import is during the network part (there was an unrelated issue where we learned AnkiWeb is not really able to indicate size) but during the import we do actually have an indication of progress, it even posts to the task listener.

Perhaps the task listener, as long as progress is being made, can extend the wake lock instead of it being a fixed value?

**Comment by david-allison-1** [comment]
> Ooo - very interesting. I like the attempt / style you're going for here though, I see what you're saying with 30min maybe working but maybe not, with the exception report we'll see I guess.
> 
> Is there anything we can bubble up to the user to help them manage the process themselves? We can't know how long an import is during the network part (there was an unrelated issue where we learned AnkiWeb is not really able to indicate size) but during the import we do actually have an indication of progress, it even posts to the task listener.
> 
> Perhaps the task listener, as long as progress is being made, can extend the wake lock instead of it being a fixed value?

I'm out of my depth here, can you extend a wakelock?

**Comment by mikehardy** [comment]
As an actual API, no, but release and reacquire with a new timeout would be the same. Not sure on how expensive that is but could be done and if expensive could be done only every x% or after x minutes or something

**Review by mikehardy — CHANGES_REQUESTED** [review]
I still don't like this one unless it actually extends on progress

To offer a concrete suggestion: perhaps in order to keep it self-contained, you can base it all on time calculated internally. If onProgressUpdate half or more of the time has elapsed, acquire a new wake lock for the duration constant defined and drop the old one

In that way as long as the task is emitting progress (which in RFC language it 'MUST DO' really if it runs that long) then it will maintain the lock

**Comment by david-allison-1** [comment]
> I still don't like this one unless it actually extends on progress
> 
> To offer a concrete suggestion: perhaps in order to keep it self-contained, you can base it all on time calculated internally. If onProgressUpdate half or more of the time has elapsed, acquire a new wake lock for the duration constant defined and drop the old one
> 
> In that way as long as the task is emitting progress (which in RFC language it 'MUST DO' really if it runs that long) then it will maintain the lock

This needs more work and effort on my behalf:

* Users have 10+GB that they're trying to import, 30m will not be viable.
* I need to heavily test the releasing of locks - maybe we'll require a "double buffering" based approach (two locks, 5 minutes apart, where only one is released)

**Comment by mikehardy** [comment]
>     * Users have 10+GB that they're trying to import, 30m will not be viable.

Yep - but it does post progress so we get a chance to extend

>     * I need to heavily test the releasing of locks - maybe we'll require a "double buffering" based approach (two locks, 5 minutes apart, where only one is released)

I'm prepared to be shocked, but I'll be shocked if grabbing a new lock before releasing the old doesn't work

**Pull request closed without merging**